stage2 x86_64: encoding helpers, fix bugs

gracefuu opened the pull request:
As stated in https://github.com/ziglang/zig/pull/8474#issuecomment-817270681, I've changed the design of this PR once. It is now just a set of helper functions for encoding, with the intention of making code more easily checkable by reading, and only _slightly_ more ergonomic to write (it helps you do bit fiddling), but otherwise doesn't try to be smart and e.g. automatically add prefixes for you.

List of bugs fixed:
1) Spilling registers to stack now correctly handles 32/64 bit. It doesn't yet handle 8 or 16 bit (#7187), which I intend to do in a separate PR making the change to the whole codegen at once, possibly at a much later date. 2ba51f87f8c8f3d0b0cc39215206bb059d898063 416b0d693340bf5d41207683c9494eebd4ce28bb
2) Typo d7158359c096fdbc1062cb5f1817a1a590340dd6
2) As a result of (1), many operations that were unnecessarily 64 bit before now correctly become 32 bit as specified, but this revealed that the specification was wrong for loading the offset to the GOT, which needed to be a 64 bit op but was specified as 32 bit. Similarly, reading function pointers from the GOT was incorrect because function pointers are 64 bit on x86_64 and aarch64, but were doing a 32 bit genSetReg instead. 91a8e8160b3580e9566870fea32af209bda13c9e

joachimschmidt557 commented on src/codegen.zig:
```diff
@@ -1487,8 +1488,14 @@ fn Function(comptime arch: std.Target.Cpu.Arch) type {
-        fn genX8664BinMath(self: *Self, inst: *ir.Inst, op_lhs: *ir.Inst, op_rhs: *ir.Inst, opx: u8, mr: u8) !MCValue {
+        fn genX8664BinMath(self: *Self, inst: *ir.Inst, op_lhs: *ir.Inst, op_rhs: *ir.Inst) !MCValue {
+            // We'll handle these ops in two steps.
+            // 1) Prepare an output register, and put one of the arguments in it
```
I may be wrong about this, but technically, the destination can be a stack location as well, I think.

gracefuu replied:
You're right. This happens when say, the LHS dies and is already on the stack, (e.g. `-0x0c(%ebp)`).

TODO: update comment to be correct.

joachimschmidt557 reviewed (commented):
Wow, nice work!

gracefuu commented on src/codegen.zig:
```diff
@@ -1549,18 +1556,109 @@ fn Function(comptime arch: std.Target.Cpu.Arch) type {
+            src: LazySrcLoc,
+            inst: Instruction,
+        ) !void {
+            inst.encodeInto(self.code) catch |err| {
```
Note to self:

```zig
inst.encodeInto(self.code) catch |err| switch (err) {
    error.OutOfMemory => return error.OutOfMemory,
    else => return self.fail(src, "Instruction.encodeInto failed because {s}", .{@errorName(err)}),
}
```
functionally does the same thing as an `if`. But `if` either lies entirely on 1 line, or will be split into 4 lines. `switch` has 1 case per line, making it only 2 lines long.

g-w1 commented on test/stage2/test.zig:
```diff
@@ -358,6 +358,81 @@ pub fn addCases(ctx: *TestContext) !void {
+            \\    return a * b;
+            \\}
+            \\
+            \\const x = mul(3, 4);
```
Since this is never referenced, it is actually never run, you probably want to do exit(x - 12); Also, I think an inline function makes execute at comptime, so this is probably redundant.

joachimschmidt557 commented:
I don't know if it's possible to add a review comment to non-visible code, but for multiplication, some code in `Sema.zig` needs to be adapted:

https://github.com/ziglang/zig/blob/faacae331d6b11c11479d96015bceb009d0e23f8/src/Sema.zig#L3807-L3812

Right now, this assumes that an operation with zero as rhs will always be the lhs, which is true for addition and subtraction, but not for multiplication as the result will be zero, not lhs, in all cases.

gracefuu commented:
I've changed approach here and replaced the monolithic instruction encoder with a set of simple helper functions because the old approach wasn't useful.

The old approach didn't make writing instructions any easier because you still had to know x86_64 to use it, and on the other hand for people who did know x86_64, it actually obscures some things such as whether the REX prefix byte is being added or not.

It was also probably quite wasteful when it comes to stack memory usage, because the Instruction struct contains every possible optional field, which probably just slows things down (not sure how well LLVM will inline them).

The new approach is just as hard to use (you still need to know x86_64) but it's much more readable. Every command corresponds exactly to zero or more bytes being written to the code array, and they occur in order.

Instead, to make writing easier, I've opted to make it more ZLS friendly by having
```
const X8664Encoder = @import("codegen/x86_64.zig").Encoder;
```
at the top of `codegen.zig`. Now tab completion works, and depending on your editor configuration might even show a description of what that particular function does.

https://user-images.githubusercontent.com/81774659/114297415-1d043200-9ae3-11eb-81bf-663b0ff6a5d3.mp4

If this push passes the test, I'll rebase onto master and mark as ready to review.

gracefuu commented:
Thanks a lot to @joachimschmidt557 for helping me debug the latest set of bugs :P

I considered whether to continue trying to write a full instruction encoder like an assembler, but decided that would be hard if we wanted it to handle the way instructions need different opcodes to handle different sizes (e.g. 0x88 for 8 bit mov, 0x89 for 16/32/64 bit mov) and different addressing forms (e.g. 0x01 for adding 16/32/64 bit from register to stack, 0x03 for adding 16/32/64 bit from stack to register).

I think the best solution currently is to proceed with the current structure. Useful functionality (e.g. genSetStack) should get abstracted into their own functions in an adhoc manner, but these functions should be written to directly emit machine code.

What should change is that each function should clearly define their contracts:

* Does it work with 8 bit?
* Can this function write TO the stack in addition to reading FROM it?
* Can this function address an offset with a large displacement (> 4GB)?
* Will exactly 1 instruction be emitted, and if not, does it use any temporary registers?

This requires that different architectures be in their own files, because the rules for x86_64 are different from arm/aarch. While working on this PR, I also found that the way different architectures have different `usingnamespace` made ZLS useless, and since roughly half of codegen.zig is arm/aarch, it was harder to navigate the codebase.

gracefuu commented:
Rebasing onto master due to LLVM 12 update

pixelherodev commented:
Nice work!

andrewrk commented:
Thanks @gracefuu! Nice work!
